### ISSUE TITLE
AbstractItemNormalizer check on Traversable for support results in false negative

### ISSUE THREAD

**Issue by renaudclosset** [api-platform/core#3461]
**API Platform version(s) affected**: v2.5.4

**Description**  
The abstract ItemNormalizer (`ApiPlatform\Core\Serializer\AbstractItemNormalizer`) checks if the passed data is an object or an instance of `\Traversable` for normalization support.

If an object is passed that implements the `\Traversable` interface, it is handled as a collection.
I presume this was added to ease the normalization of (Doctrine) Collection objects.

**How to reproduce**  
Create an object implementing the `\Traversable` interface with a custom DataTransformer. The DataTransformer will never be called.

**Possible Solution**  
Don't check on `\Traversable` but rely on the fact if this class is a Resource.

```php
public function supportsNormalization($data, $format = null)
{
  if (!\is_object($data)) {
    return false;
  }
  return $this->resourceClassResolver->isResourceClass($this->getObjectClass($data));
}
```

**Comment by teohhanhui** [comment]
Basically, we treat `iterable` as a collection. It's not specific to Doctrine `Collection`.

Your proposed alternative is not feasible.

**Comment by teohhanhui** [comment]
Hmm... Even if we make the change as you've suggested, you'd still run into an issue here:

https://github.com/api-platform/core/blob/48fc5bb75e90d7ccef4ba540630a85bd12016f44/src/Api/ResourceClassResolver.php#L48

May we recommend not implementing `\Traversable` in your resource class?

**Comment by renaudclosset** [comment]
As far as i could tell the `getResourceClass` is never called in this scope, but it is based on the `isResourceClass` method.
The `isResourceClass` looks directly into the `ResourceClassResolver::$localIsResourceClassCache` list, which in turn is fed by the class' `resourceNameCollectionFactory`, which lists all classes configured as Resources.

This issue is actually the question "_Are we dealing with an Item or a Collection?_"
If we base this on the implementation `\Traversable` in a class, this could results if false negatives. But looking a the Resource list that was explicitly defined by the user seems to be a more robust approach.

If some Resource are automatically added, without explicit user configuration, we could fallback to the `\Traversable` implementation check. no?

**Comment by renaudclosset** [comment]
Hey,

I created a pull request that should deal with this issue without any BC breaks.
Could that be reviewed/discussed?

Thanks

**Comment by teohhanhui** [comment]
Hi, the problem is not whether `getResourceClass` is called in this context. The problem is that even if the code works in this part, it'll not work elsewhere. Using a resource class which implements `\Traversable` is simply not supported. I don't see a feasible solution, do you?

**Comment by teohhanhui** [comment]
Let's keep the discussion here.

**Comment by renaudclosset** [comment]
I understand your concern, but the change proposed here only affects the result of classes defined/configured as a Resource (in the configuration). All other classes that are not defined as Resource (e.g. collections, as these will never be defined as an Resource/Item in the configuration) will still be checked by the `\Traversable` statement.

So the only possible issue I could identify here is if somebody marked a class as an `@ApiResource` but expect it to be normalized as a Collection of Resources. But this would not make any sense.

Or are there any cases where a Resource should be normalized as a Collection? (_as I am not yet very familiar with the inner workings of this library_)

I looked through the code of the "api-plaform/core" library, and the check for `\Traversable` only happens on the two places we identified here.
